### ISSUE TITLE
[FEATURE] add ability to scan npm / scan targets

### ISSUE THREAD

**Issue by zombieprocess** [sonatype-nexus-community/scan-gradle-plugin#139]
* **What are you trying to do?**

Hi all, I have a multi-module project - Angular Frontend (npm) and a Java backend. 

I'm trying to scan the whole project with `./gradlew nexusIQScan` and get the same finding results I do as my Jenkins pipeline using the [Jenkins plugin](https://help.sonatype.com/iqserver/analysis/npm-application-analysis#npmApplicationAnalysis-StepstoanalyzeusingtheJenkinsplugin) with the `iqScanPAtterns` targeting the frontend `packages-lock.json` and the backend built `application.jar`.

When using `2.6.0` of `scan-gradle-plugin` I only get results from the backend java module. No findings/results are returned for the frontend. (npm)

* **What feature or behavior is this required for?**

This would allow developers to scan their project **locally** before doing a `git push`  and get a consistent scan result that they are seeing in the CI pipeline. This would allow them to fix any findings before the code changes are pushed to the `git` repo.

* **How could we solve this issue? (Not knowing is okay!)**

Add the ability to explicitly target **files** in the `Scan configuration properties` or implement `Scan Patterns` like the Jenkins plugin:

Mocking the config in `build.gradle` :
**Example:**
```
nexusIQScan {
    username = 'admin' // Make sure to use an user with the role 'Application Evaluator' in the given IQ Server application
    password = 'pass'
    serverUrl = 'http://localhost:8070'
    applicationId = 'app'
    iqScanPatterns = '**/frontend/packages-lock.json,**/backend/build/app-*.jar'
}
```
`iqScanPatterns` is what the Jenkins plugin uses to target npm packages via a shrinkwrap or packages-lock.json files. (direct/transitive)

```
> Task :nexusIQScan
... Scanning application...
Discovered commit hash 'hash' via jGit
Scan target: /home/user/app/backend/build/app-1.0.0.jar
Scan target: /home/user/app/frontend/package-lock.json
Scan configuration properties:
   dirExcludes=null
   dirIncludes=null
   fileExcludes=
   fileIncludes=
 Started scanning target: ....
```
This is how my output looks from a jenkins pipeline, which is nearly identical to the `nexusIQScan` task output, sans the "Scan Target".

* **Anything else?**

cc @bhamail / @DarthHater / @guillermo-varela / @shaikhu

**Comment by guillermo-varela** [comment]
Hi @zombieprocess!

I'm leaving some notes from my initial findings:

The Jenkins plugin takes the `iqScanPatterns` paths and for each one collects the files matching the pattern into a collection `targets`:

https://github.com/jenkinsci/nexus-platform-plugin/blob/e8c2e705cc5c7699611c22cb293ef3ed8aa27140/src/main/java/org/sonatype/nexus/ci/iq/RemoteScanner.groovy#L88-L95

Those target files are sent to the scanner using the method `scan`
https://github.com/jenkinsci/nexus-platform-plugin/blob/e8c2e705cc5c7699611c22cb293ef3ed8aa27140/src/main/java/org/sonatype/nexus/ci/iq/RemoteScanner.groovy#L97

Although this plugin uses an overloaded version of that `scan` method, it has the `targets` parameter available and currently we're sending an empty List, that because we're not scanning all JAR (AAR or any other format) binary file under the working directory but instead we're collecting the resolved dependencies under `modules`:

https://github.com/sonatype-nexus-community/scan-gradle-plugin/blob/8b2e19fc9af9bfb44519f3707ba7885295192e9c/src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/scan/NexusIqScanTask.java#L110-L111

In theory, we could send both a list of target files (which would need to be resolved from the given patterns in this plugin) and the modules with the dependencies to the scanner, but that requires some tests to check we don't get any undesired side-effects.

**Comment by zombieprocess** [comment]
Thanks for getting back to me @guillermo-varela!  

What you outline in your investigation of the `nexus-platform-plugin` makes sense to me. I'd love to see a change where we can populate the `targets` list from a config object in the `build.gradle` or similar. 

If I can help test anything against my project, let me know. 😀

**Comment by zombieprocess** [comment]
Hi @guillermo-varela!

Just following up on this, as things have changed in the jenkins nexus scanning environment recently making this enhancement/feature request more important to my team. (maybe others?)

As you can see here the `nexus-jenkins-plugin` has been delisted (links below) and we would like to convert to CLI `scan-gradle-plugin` (or maven)  scanning of our NPM/Node JS projects.
- https://plugins.jenkins.io/nexus-jenkins-plugin/
- https://github.com/jenkins-infra/helpdesk/issues/3742

Right now the only way we can reliably scan our `package-lock.json` is via the `nexus-jenkins-plugin` which as far as we can tell is not easily available anymore. If you have any alternative ideas, I'm all ears. 👂 

If you'd like we can reach out to our contacts at Sonatype if that would help prioritize and fund any updates.

**Comment by guillermo-varela** [comment]
Hi @zombieprocess!

It's much appreciated that you make a case for this feature request 😄 
While I cannot commit to a date, this is something that will be worked on based on your comments.

Also, as a temporary measure, the `hpi` file for latest version of the Jenkins plugin is available on our Downloads site so the plugin can be installed/updated manually:
https://help.sonatype.com/iqserver/product-information/download-and-compatibility#DownloadandCompatibility-Jenkins

Hopefully that works while this feature is developed.

Thanks.

**Comment by guillermo-varela** [comment]
Hi again @zombieprocess!

As an update on this, I was about to start working on this feature but hit a wall related to a library this plugin depends on: https://github.com/sonatype-nexus-community/scan-gradle-plugin/issues/146

To avoid breaking backwards compatibility just yet, I'll do my best to develop this feature using the libraries the plugin currently has.

Thanks and stay tuned!

**Comment by zombieprocess** [comment]
Great, thank you! I appreciate the updates

**Comment by guillermo-varela** [comment]
Hi @zombieprocess!

One more update: significant progress was made and there is now a draft PR
https://github.com/sonatype-nexus-community/scan-gradle-plugin/pull/150

Once we have unit/integration tests the PR will be ready for review.

**Comment by zombieprocess** [comment]
Great, thank you @guillermo-varela !! Please do let me know when it is ready. 👍

**Comment by guillermo-varela** [comment]
Hi @zombieprocess!

This feature has been included in our latest release 2.7.0:
https://github.com/sonatype-nexus-community/scan-gradle-plugin/releases/tag/2.7.0

Please let us know if this meets the expectations on this issue.

**Comment by zombieprocess** [comment]
Hi @guillermo-varela - I tested the feature and it looks great! Thank you for tackling this. Our scans now report our frontend and backend scan results in a single report, which is exactly what we needed. Thank you again! 🙌
